### ISSUE TITLE
Crash on malformed input

### ISSUE THREAD

**Issue by Shnatsel** [RustAudio/lewton#32]
When given a malformed input lewton attempts to allocate enormous amounts of memory and crashes. Testcase: [minimized_crash_testcase.ogg.gz](https://github.com/RustAudio/lewton/files/2361339/minimized_crash_testcase.ogg.gz) (gzipped so that github would accept the upload)

Command to reproduce the crash: 
`RUSTFLAGS='--cfg=fuzzing' cargo run --release --example perf /path/to/testcase` 
The output will be:
```
memory allocation of 4611686018427387908 bytes failed
fish: “env RUSTFLAGS='--cfg=fuzzing' c…” terminated by signal SIGABRT (Abort)
```

When run in debug mode with command 
`RUSTFLAGS='--cfg=fuzzing' cargo run --example perf /path/to/testcase` the error message is:
```
thread 'main' panicked at 'attempt to subtract with overflow', src/audio.rs:639:18
```

Found using [afl.rs](https://github.com/rust-fuzz/afl.rs)

**Comment by est31** [comment]
Thanks for the next report!

[Repacked](https://github.com/RustAudio/ogg/commit/f1fcf4bf74d99a86885c3793b3e295ea058d2564) version of the test case: [32_minimized_crash_testcase.ogg.zip](https://github.com/RustAudio/lewton/files/2363080/32_minimized_crash_testcase.ogg.zip)

(needs no `--cfg=fuzzing` as the CRC sums are fixed)

**Comment by est31** [comment]
Note: I don't consider large allocations a bug. I mean how can you distinguish a legitimate large allocation from an illegitimate one? The problem is solvable far more nicely with failible allocators. Hopefully Rust will get support one day. Until then there has to be a panic.

*However* I do consider the overflow in debug more to be a bug. Thus I'll push a fix soon.

**Comment by est31** [comment]
Fixed in commit b20e267fbf92e8efedbc240767b79d4f7a159633

**Comment by Shnatsel** [comment]
> I don't consider large allocations a bug. I mean how can you distinguish a legitimate large allocation from an illegitimate one?

This is usually done by limiting the amount of allocated memory to some sane default, and letting people override it if they're really dealing with enormous amounts of data. In Rust we can easily allow the API user to override these limits via the [builder pattern](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html).

See https://libpng.sourceforge.io/decompression_bombs.html for more info on how a similar issue was solved in libpng. See also the [Limits](https://docs.rs/flif/0.3.0/flif/struct.Limits.html) struct from flif crate.
